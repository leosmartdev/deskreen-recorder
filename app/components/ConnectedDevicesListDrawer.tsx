/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable react/destructuring-assignment */
import { remote } from 'electron';
import React, { useEffect, useState, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import {
  Button,
  Text,
  Position,
  Drawer,
  Card,
  Alert,
  H4,
} from '@blueprintjs/core';
import { Row, Col } from 'react-flexbox-grid';
import { createStyles, makeStyles } from '@material-ui/core/styles';
import CloseOverlayButton from './CloseOverlayButton';
import ConnectedDevicesService from '../features/ConnectedDevicesService';
import SharingSessionService from '../features/SharingSessionService';
import DeviceInfoCallout from './DeviceInfoCallout';
import SharingSourcePreviewCard from './SharingSourcePreviewCard';
import isWithReactRevealAnimations from '../utils/isWithReactRevealAnimations';
import isProduction from '../utils/isProduction';

const sharingSessionService = remote.getGlobal(
  'sharingSessionService'
) as SharingSessionService;
const connectedDevicesService = remote.getGlobal(
  'connectedDevicesService'
) as ConnectedDevicesService;

const Fade = require('react-reveal/Fade');

const disconnectPeerAndDestroySharingSessionBySessionID = (
  sharingSessionID: string
) => {
  const sharingSession = sharingSessionService.sharingSessions.get(
    sharingSessionID
  );
  sharingSession?.disconnectByHostMachineUser();
  sharingSession?.destroy();
  sharingSessionService.sharingSessions.delete(sharingSessionID);
};

interface ConnectedDevicesListDrawerProps {
  isOpen: boolean;
  handleToggle: () => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  stepperRef: any;
}

const useStyles = makeStyles(() =>
  createStyles({
    drawerRoot: { overflowY: 'scroll', overflowX: 'hidden' },
    drawerInnerTopPanel: { padding: '20px 10px 0px 30px' },
    connectedDevicesRoot: { padding: '10px 20px' },
    topHeader: {
      marginRight: '20px',
      fontSize: '20px',
      fontWeight: 900,
    },
    zoomFullWidth: {
      width: '100%',
    },
  })
);

export default function ConnectedDevicesListDrawer(
  props: ConnectedDevicesListDrawerProps
) {
  const { t } = useTranslation();
  const classes = useStyles();

  const [isAlertDisconectAllOpen, setIsAlertDisconectAllOpen] = useState(false);

  const [devicesDisplayed, setDevicesDisplayed] = useState(new Map());

  useEffect(() => {
    const map = new Map();
    connectedDevicesService.getDevices().forEach((el) => {
      map.set(el.id, true);
    });
    setDevicesDisplayed(map);
  }, [setDevicesDisplayed]);

  const handleDisconnectOneDevice = useCallback(async (id: string) => {
    const device = connectedDevicesService.devices.find(
      (d: Device) => d.id === id
    );
    if (!device) return;
    disconnectPeerAndDestroySharingSessionBySessionID(device.sharingSessionID);
    connectedDevicesService.removeDeviceByID(id);
  }, []);

  const handleDisconnectAll = useCallback(() => {
    connectedDevicesService.devices.forEach((device: Device) => {
      disconnectPeerAndDestroySharingSessionBySessionID(
        device.sharingSessionID
      );
    });
    connectedDevicesService.removeAllDevices();
  }, []);

  const hideOneDeviceInDevicesDisplayed = useCallback(
    (id) => {
      const newDevicesDisplayed = new Map(devicesDisplayed);
      newDevicesDisplayed.set(id, false);
      setDevicesDisplayed(newDevicesDisplayed);
    },
    [devicesDisplayed, setDevicesDisplayed]
  );

  const hideAllDevicesInDevicesDisplayed = useCallback(() => {
    const newDevicesDisplayed = new Map(devicesDisplayed);
    [...newDevicesDisplayed.keys()].forEach((key) => {
      newDevicesDisplayed.set(key, false);
    });
    setDevicesDisplayed(newDevicesDisplayed);
  }, [devicesDisplayed, setDevicesDisplayed]);

  const handleDisconnectAndHideOneDevice = useCallback(
    (id) => {
      setTimeout(
        async () => {
          await handleDisconnectOneDevice(id);
          hideOneDeviceInDevicesDisplayed(id);
        },
        isProduction() ? 1000 : 0
      );
    },
    [handleDisconnectOneDevice, hideOneDeviceInDevicesDisplayed]
  );

  const handleDisconnectAndHideAllDevices = useCallback(() => {
    hideAllDevicesInDevicesDisplayed();
    setTimeout(
      () => {
        handleDisconnectAll();
        props.handleToggle();
        props.stepperRef.current.handleReset();
      },
      isProduction() ? 1000 : 0
    );
  }, [handleDisconnectAll, hideAllDevicesInDevicesDisplayed, props]);

  const disconnectAllCancelButtonText = t('No, Cancel');
  const disconnectAllConfirmButtonText = t('Yes, Disconnect All');

  return (
    <>
      <Drawer
        className={classes.drawerRoot}
        position={Position.BOTTOM}
        size={Drawer.SIZE_LARGE}
        isOpen={props.isOpen}
        onClose={props.handleToggle}
        transitionDuration={isWithReactRevealAnimations() ? 700 : 0}
        // transitionDuration={0}
      >
        <Row between="xs" middle="xs" className={classes.drawerInnerTopPanel}>
          <Col xs={11}>
            <Row middle="xs">
              <div className={classes.topHeader}>
                <Text className="bp3-text-muted">{t('Connected Devices')}</Text>
              </div>
              <Button
                intent="danger"
                disabled={connectedDevicesService.getDevices().length === 0}
                onClick={() => {
                  setIsAlertDisconectAllOpen(true);
                }}
                icon="disable"
                style={{
                  borderRadius: '100px',
                }}
              >
                {t('Disconnect all devices')}
              </Button>
            </Row>
          </Col>
          <Col xs={1}>
            <CloseOverlayButton onClick={props.handleToggle} isDefaultStyles />
          </Col>
        </Row>
        <Row className={classes.connectedDevicesRoot}>
          <Col xs={12}>
            <Fade
              bottom
              cascade
              duration={isWithReactRevealAnimations() ? 700 : 0}
            >
              <div className={classes.zoomFullWidth}>
                {connectedDevicesService.getDevices().map((device) => {
                  return (
                    <div key={device.id}>
                      <Fade
                        collapse
                        opposite
                        when={devicesDisplayed.get(device.id)}
                        duration={isProduction() ? 700 : 0}
                      >
                        <Card className="connected-device-card">
                          <Row middle="xs">
                            <Col xs={6}>
                              <DeviceInfoCallout
                                deviceType={device.deviceType}
                                deviceOS={device.deviceOS}
                                deviceIP={device.deviceIP}
                                sharingSessionID={device.sharingSessionID}
                                deviceBrowser={device.deviceBrowser}
                              />
                            </Col>
                            <Col xs={6}>
                              <SharingSourcePreviewCard
                                sharingSourceID={
                                  sharingSessionService.sharingSessions.get(
                                    device.sharingSessionID
                                  )?.desktopCapturerSourceID
                                }
                              />
                            </Col>
                          </Row>
                          <Row center="xs">
                            <Button
                              id={`disconnect-device-${device.deviceIP}`}
                              intent="danger"
                              onClick={(): void => {
                                handleDisconnectAndHideOneDevice(device.id);
                              }}
                              icon="disable"
                              style={{
                                borderRadius: '100px',
                              }}
                            >
                              {t('Disconnect')}
                            </Button>
                          </Row>
                        </Card>
                      </Fade>
                    </div>
                  );
                })}
              </div>
            </Fade>
          </Col>
        </Row>
      </Drawer>
      <Alert
        isOpen={isAlertDisconectAllOpen}
        onClose={() => {
          setIsAlertDisconectAllOpen(false);
        }}
        icon="warning-sign"
        cancelButtonText={disconnectAllCancelButtonText}
        confirmButtonText={disconnectAllConfirmButtonText}
        intent="danger"
        canEscapeKeyCancel
        canOutsideClickCancel
        onCancel={() => {
          setIsAlertDisconectAllOpen(false);
        }}
        onConfirm={handleDisconnectAndHideAllDevices}
        transitionDuration={isWithReactRevealAnimations() ? 700 : 0}
      >
        <H4>
          {t(
            'Are you sure you want to disconnect all connected viewing devices?'
          )}
        </H4>
        <Text>{`${t('This step can not be undone')}.`}</Text>
        <Text>
          {`${t('You will have to connect all devices manually again')}.`}
        </Text>
      </Alert>
    </>
  );
}
