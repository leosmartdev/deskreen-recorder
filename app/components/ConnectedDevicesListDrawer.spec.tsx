import React from 'react';
import Enzyme, { mount } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import ConnectedDevicesListDrawer from './ConnectedDevicesListDrawer';

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();

jest.mock('electron', () => {
  return {
    remote: {
      getGlobal: (globalName: string) => {
        if (globalName === 'connectedDevicesService') {
          return {
            getDevices: () => [],
          };
        }
        return {};
      },
    },
  };
});

it('should match exact snapshot', () => {
  const subject = mount(
    <>
      <Router>
        <ConnectedDevicesListDrawer
          isOpen
          handleToggle={() => {}}
          stepperRef={null}
        />
      </Router>
    </>
  );
  expect(EnzymeToJson(subject)).toMatchSnapshot();
});
