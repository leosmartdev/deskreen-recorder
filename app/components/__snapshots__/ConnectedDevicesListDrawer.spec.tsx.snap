// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match exact snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ConnectedDevicesListDrawer
      handleToggle={[Function]}
      isOpen={true}
      stepperRef={null}
    >
      <Blueprint3.Drawer
        canOutsideClickClose={true}
        className="makeStyles-drawerRoot-1"
        isOpen={true}
        onClose={[Function]}
        position="bottom"
        size="90%"
        style={Object {}}
        transitionDuration={0}
        vertical={false}
      >
        <Blueprint3.Overlay
          autoFocus={true}
          backdropProps={Object {}}
          canEscapeKeyClose={true}
          canOutsideClickClose={true}
          className="bp3-overlay-container"
          enforceFocus={true}
          hasBackdrop={true}
          isOpen={true}
          lazy={true}
          onClose={[Function]}
          position="bottom"
          size="90%"
          style={Object {}}
          transitionDuration={0}
          transitionName="bp3-overlay"
          usePortal={true}
          vertical={false}
        >
          <Blueprint3.Portal
            container={
              <body
                class="bp3-overlay-open"
              >
                <div
                  class="bp3-portal"
                >
                  <div
                    class="bp3-overlay bp3-overlay-open bp3-overlay-container"
                  >
                    <div
                      class="bp3-overlay-backdrop bp3-overlay-appear bp3-overlay-appear-active"
                      tabindex="0"
                    />
                    <div
                      class="bp3-drawer bp3-position-bottom makeStyles-drawerRoot-1 bp3-overlay-content bp3-overlay-appear bp3-overlay-appear-active"
                      style="height: 90%;"
                      tabindex="0"
                    >
                      <div
                        class="makeStyles-drawerInnerTopPanel-2 row middle-xs between-xs"
                      >
                        <div
                          class="col-xs-11"
                        >
                          <div
                            class="row middle-xs"
                          >
                            <div
                              class="makeStyles-topHeader-4"
                            >
                              <div
                                class="bp3-text-muted"
                              />
                            </div>
                            <button
                              class="bp3-button bp3-disabled bp3-intent-danger"
                              disabled=""
                              style="border-radius: 100px;"
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                class="bp3-icon bp3-icon-disable"
                                icon="disable"
                              >
                                <svg
                                  data-icon="disable"
                                  height="16"
                                  viewBox="0 0 16 16"
                                  width="16"
                                >
                                  <desc>
                                    disable
                                  </desc>
                                  <path
                                    d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm-6 8c0-3.31 2.69-6 6-6 1.3 0 2.49.42 3.47 1.12l-8.35 8.35c-.7-.98-1.12-2.17-1.12-3.47zm6 6c-1.3 0-2.49-.42-3.47-1.12l8.35-8.35c.7.98 1.12 2.17 1.12 3.47 0 3.32-2.68 6-6 6z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          class="col-xs-1"
                        >
                          <button
                            class="bp3-button makeStyles-closeButton-6 undefined"
                            id="close-overlay-button"
                            type="button"
                          >
                            <span
                              class="bp3-button-text"
                            >
                              <span
                                class="bp3-icon bp3-icon-cross"
                                icon="cross"
                              >
                                <svg
                                  data-icon="cross"
                                  height="30"
                                  viewBox="0 0 20 20"
                                  width="30"
                                >
                                  <desc>
                                    cross
                                  </desc>
                                  <path
                                    d="M11.41 10l4.29-4.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-4.29-4.3a1.003 1.003 0 00-1.42 1.42L8.59 10 4.3 14.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4.29-4.3 4.29 4.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                            </span>
                          </button>
                        </div>
                      </div>
                      <div
                        class="makeStyles-connectedDevicesRoot-3 row"
                      >
                        <div
                          class="col-xs-12"
                        >
                          <div
                            class="react-reveal makeStyles-zoomFullWidth-5"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </body>
            }
          >
            <Portal
              containerInfo={
                <div
                  class="bp3-portal"
                >
                  <div
                    class="bp3-overlay bp3-overlay-open bp3-overlay-container"
                  >
                    <div
                      class="bp3-overlay-backdrop bp3-overlay-appear bp3-overlay-appear-active"
                      tabindex="0"
                    />
                    <div
                      class="bp3-drawer bp3-position-bottom makeStyles-drawerRoot-1 bp3-overlay-content bp3-overlay-appear bp3-overlay-appear-active"
                      style="height: 90%;"
                      tabindex="0"
                    >
                      <div
                        class="makeStyles-drawerInnerTopPanel-2 row middle-xs between-xs"
                      >
                        <div
                          class="col-xs-11"
                        >
                          <div
                            class="row middle-xs"
                          >
                            <div
                              class="makeStyles-topHeader-4"
                            >
                              <div
                                class="bp3-text-muted"
                              />
                            </div>
                            <button
                              class="bp3-button bp3-disabled bp3-intent-danger"
                              disabled=""
                              style="border-radius: 100px;"
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                class="bp3-icon bp3-icon-disable"
                                icon="disable"
                              >
                                <svg
                                  data-icon="disable"
                                  height="16"
                                  viewBox="0 0 16 16"
                                  width="16"
                                >
                                  <desc>
                                    disable
                                  </desc>
                                  <path
                                    d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm-6 8c0-3.31 2.69-6 6-6 1.3 0 2.49.42 3.47 1.12l-8.35 8.35c-.7-.98-1.12-2.17-1.12-3.47zm6 6c-1.3 0-2.49-.42-3.47-1.12l8.35-8.35c.7.98 1.12 2.17 1.12 3.47 0 3.32-2.68 6-6 6z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          class="col-xs-1"
                        >
                          <button
                            class="bp3-button makeStyles-closeButton-6 undefined"
                            id="close-overlay-button"
                            type="button"
                          >
                            <span
                              class="bp3-button-text"
                            >
                              <span
                                class="bp3-icon bp3-icon-cross"
                                icon="cross"
                              >
                                <svg
                                  data-icon="cross"
                                  height="30"
                                  viewBox="0 0 20 20"
                                  width="30"
                                >
                                  <desc>
                                    cross
                                  </desc>
                                  <path
                                    d="M11.41 10l4.29-4.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-4.29-4.3a1.003 1.003 0 00-1.42 1.42L8.59 10 4.3 14.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4.29-4.3 4.29 4.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </span>
                            </span>
                          </button>
                        </div>
                      </div>
                      <div
                        class="makeStyles-connectedDevicesRoot-3 row"
                      >
                        <div
                          class="col-xs-12"
                        >
                          <div
                            class="react-reveal makeStyles-zoomFullWidth-5"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              }
            >
              <TransitionGroup
                appear={true}
                childFactory={[Function]}
                className="bp3-overlay bp3-overlay-open bp3-overlay-container"
                component="div"
                onKeyDown={[Function]}
              >
                <div
                  className="bp3-overlay bp3-overlay-open bp3-overlay-container"
                  onKeyDown={[Function]}
                >
                  <CSSTransition
                    addEndListener={[Function]}
                    appear={true}
                    classNames="bp3-overlay"
                    in={true}
                    key=".$__backdrop"
                    onExited={[Function]}
                    timeout={0}
                  >
                    <Transition
                      addEndListener={[Function]}
                      appear={true}
                      enter={true}
                      exit={true}
                      in={true}
                      mountOnEnter={false}
                      onEnter={[Function]}
                      onEntered={[Function]}
                      onEntering={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onExiting={[Function]}
                      timeout={0}
                      unmountOnExit={false}
                    >
                      <div
                        className="bp3-overlay-backdrop"
                        onMouseDown={[Function]}
                        tabIndex={0}
                      />
                    </Transition>
                  </CSSTransition>
                  <CSSTransition
                    addEndListener={[Function]}
                    appear={true}
                    classNames="bp3-overlay"
                    in={true}
                    key=".$.0"
                    onExited={[Function]}
                    timeout={0}
                  >
                    <Transition
                      addEndListener={[Function]}
                      appear={true}
                      enter={true}
                      exit={true}
                      in={true}
                      mountOnEnter={false}
                      onEnter={[Function]}
                      onEntered={[Function]}
                      onEntering={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onExiting={[Function]}
                      timeout={0}
                      unmountOnExit={false}
                    >
                      <div
                        className="bp3-drawer bp3-position-bottom makeStyles-drawerRoot-1 bp3-overlay-content"
                        style={
                          Object {
                            "height": "90%",
                          }
                        }
                        tabIndex={0}
                      >
                        <Row
                          between="xs"
                          className="makeStyles-drawerInnerTopPanel-2"
                          middle="xs"
                        >
                          <div
                            className="makeStyles-drawerInnerTopPanel-2 row middle-xs between-xs"
                          >
                            <Col
                              xs={11}
                            >
                              <div
                                className="col-xs-11"
                              >
                                <Row
                                  middle="xs"
                                >
                                  <div
                                    className="row middle-xs"
                                  >
                                    <div
                                      className="makeStyles-topHeader-4"
                                    >
                                      <Blueprint3.Text
                                        className="bp3-text-muted"
                                        ellipsize={false}
                                        tagName="div"
                                      >
                                        <div
                                          className="bp3-text-muted"
                                        />
                                      </Blueprint3.Text>
                                    </div>
                                    <Blueprint3.Button
                                      disabled={true}
                                      icon="disable"
                                      intent="danger"
                                      onClick={[Function]}
                                      style={
                                        Object {
                                          "borderRadius": "100px",
                                        }
                                      }
                                    >
                                      <button
                                        className="bp3-button bp3-disabled bp3-intent-danger"
                                        disabled={true}
                                        onBlur={[Function]}
                                        onKeyDown={[Function]}
                                        onKeyUp={[Function]}
                                        style={
                                          Object {
                                            "borderRadius": "100px",
                                          }
                                        }
                                        tabIndex={-1}
                                        type="button"
                                      >
                                        <Blueprint3.Icon
                                          icon="disable"
                                          key="leftIcon"
                                        >
                                          <span
                                            className="bp3-icon bp3-icon-disable"
                                            icon="disable"
                                          >
                                            <svg
                                              data-icon="disable"
                                              height={16}
                                              viewBox="0 0 16 16"
                                              width={16}
                                            >
                                              <desc>
                                                disable
                                              </desc>
                                              <path
                                                d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm-6 8c0-3.31 2.69-6 6-6 1.3 0 2.49.42 3.47 1.12l-8.35 8.35c-.7-.98-1.12-2.17-1.12-3.47zm6 6c-1.3 0-2.49-.42-3.47-1.12l8.35-8.35c.7.98 1.12 2.17 1.12 3.47 0 3.32-2.68 6-6 6z"
                                                fillRule="evenodd"
                                                key="0"
                                              />
                                            </svg>
                                          </span>
                                        </Blueprint3.Icon>
                                        <Blueprint3.Icon
                                          key="rightIcon"
                                        />
                                      </button>
                                    </Blueprint3.Button>
                                  </div>
                                </Row>
                              </div>
                            </Col>
                            <Col
                              xs={1}
                            >
                              <div
                                className="col-xs-1"
                              >
                                <CloseOverlayButton
                                  isDefaultStyles={true}
                                  onClick={[Function]}
                                >
                                  <Blueprint3.Button
                                    className="makeStyles-closeButton-6 undefined"
                                    id="close-overlay-button"
                                    onClick={[Function]}
                                  >
                                    <button
                                      className="bp3-button makeStyles-closeButton-6 undefined"
                                      id="close-overlay-button"
                                      onBlur={[Function]}
                                      onClick={[Function]}
                                      onKeyDown={[Function]}
                                      onKeyUp={[Function]}
                                      type="button"
                                    >
                                      <Blueprint3.Icon
                                        key="leftIcon"
                                      />
                                      <span
                                        className="bp3-button-text"
                                        key="text"
                                      >
                                        <Blueprint3.Icon
                                          icon="cross"
                                          iconSize={30}
                                        >
                                          <span
                                            className="bp3-icon bp3-icon-cross"
                                            icon="cross"
                                          >
                                            <svg
                                              data-icon="cross"
                                              height={30}
                                              viewBox="0 0 20 20"
                                              width={30}
                                            >
                                              <desc>
                                                cross
                                              </desc>
                                              <path
                                                d="M11.41 10l4.29-4.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-4.29-4.3a1.003 1.003 0 00-1.42 1.42L8.59 10 4.3 14.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4.29-4.3 4.29 4.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z"
                                                fillRule="evenodd"
                                                key="0"
                                              />
                                            </svg>
                                          </span>
                                        </Blueprint3.Icon>
                                      </span>
                                      <Blueprint3.Icon
                                        key="rightIcon"
                                      />
                                    </button>
                                  </Blueprint3.Button>
                                </CloseOverlayButton>
                              </div>
                            </Col>
                          </div>
                        </Row>
                        <Row
                          className="makeStyles-connectedDevicesRoot-3"
                        >
                          <div
                            className="makeStyles-connectedDevicesRoot-3 row"
                          >
                            <Col
                              xs={12}
                            >
                              <div
                                className="col-xs-12"
                              >
                                <Fade
                                  bottom={true}
                                  cascade={true}
                                  duration={0}
                                >
                                  <RevealBase
                                    bottom={true}
                                    cascade={true}
                                    fraction={0.2}
                                    inEffect={
                                      Object {
                                        "count": 1,
                                        "delay": 0,
                                        "duration": 0,
                                        "forever": undefined,
                                        "make": [Function],
                                        "reverse": undefined,
                                        "style": Object {
                                          "animationFillMode": "both",
                                        },
                                      }
                                    }
                                    outEffect={
                                      Object {
                                        "count": 1,
                                        "delay": 0,
                                        "duration": 0,
                                        "forever": undefined,
                                        "make": [Function],
                                        "reverse": undefined,
                                        "style": Object {
                                          "animationFillMode": "both",
                                        },
                                      }
                                    }
                                    refProp="ref"
                                  >
                                    <div
                                      className="react-reveal makeStyles-zoomFullWidth-5"
                                      style={
                                        Object {
                                          "opacity": undefined,
                                        }
                                      }
                                    />
                                  </RevealBase>
                                </Fade>
                              </div>
                            </Col>
                          </div>
                        </Row>
                      </div>
                    </Transition>
                  </CSSTransition>
                </div>
              </TransitionGroup>
            </Portal>
          </Blueprint3.Portal>
        </Blueprint3.Overlay>
      </Blueprint3.Drawer>
      <Blueprint3.Alert
        canEscapeKeyCancel={true}
        canOutsideClickCancel={true}
        cancelButtonText=""
        confirmButtonText=""
        icon="warning-sign"
        intent="danger"
        isOpen={false}
        onCancel={[Function]}
        onClose={[Function]}
        onConfirm={[Function]}
        transitionDuration={0}
      >
        <Blueprint3.Dialog
          canEscapeKeyClose={true}
          canOutsideClickClose={true}
          className="bp3-alert"
          isOpen={false}
          onCancel={[Function]}
          onClose={[Function]}
          onConfirm={[Function]}
          transitionDuration={0}
        >
          <Blueprint3.Overlay
            autoFocus={true}
            backdropProps={Object {}}
            canEscapeKeyClose={true}
            canOutsideClickClose={true}
            className="bp3-overlay-scroll-container"
            enforceFocus={true}
            hasBackdrop={true}
            isOpen={false}
            lazy={true}
            onCancel={[Function]}
            onClose={[Function]}
            onConfirm={[Function]}
            transitionDuration={0}
            transitionName="bp3-overlay"
            usePortal={true}
          />
        </Blueprint3.Dialog>
      </Blueprint3.Alert>
    </ConnectedDevicesListDrawer>
  </Router>
</BrowserRouter>
`;
